lib/transfer.c

```diff
@@ -231,10 +231,16 @@ CURLcode Curl_fillreadbuffer(struct Curl_easy *data, size_t bytes,
     extra_data = data->state.in;
   }
 
-  Curl_set_in_callback(data, true);
-  nread = readfunc(data->req.upload_fromhere, 1,
-                   buffersize, extra_data);
-  Curl_set_in_callback(data, false);
+  if(!data->req.fread_func_eof) {
+    Curl_set_in_callback(data, true);
+    nread = readfunc(data->req.upload_fromhere, 1,
+                     buffersize, extra_data);
+    Curl_set_in_callback(data, false);
+    /* make sure the callback is not called again after EOF */
+    data->req.fread_func_eof = (nread != buffersize);
```

Comment:
that is going to set it for situations other than eof, i'm curious why you didn't do `data->req.fread_func_eof = !nread;`. like what if there's a pause for example

Reply:
Ah right, pause.... 

Well I did it of course because returning an error should not call it again either.

Reply:
it's not just that there are other scenarios like libcurl requests 100 bytes and 50 bytes or something is returned